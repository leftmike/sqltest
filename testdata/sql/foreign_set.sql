--
-- Test FOREIGN KEY constraints
--
-- {{if eq Dialect "sqlite3"}}{{Skip}}{{end}}

DROP TABLE IF EXISTS tbl3;

DROP TABLE IF EXISTS tbl2;

DROP TABLE IF EXISTS tbl1;

CREATE TABLE tbl1 (
    c1 int PRIMARY KEY,
    c2 int,
    c3 int
);

CREATE TABLE tbl2 (
    c4 int PRIMARY KEY,
    c5 int,
    c6 int REFERENCES tbl1 ON DELETE SET DEFAULT ON UPDATE SET NULL DEFAULT 1
);

CREATE TABLE tbl3 (
    c4 int PRIMARY KEY,
    c5 int,
    c6 int REFERENCES tbl1 ON DELETE SET NULL ON UPDATE SET DEFAULT DEFAULT 4 NOT NULL
);

INSERT INTO tbl1 VALUES
    (1, 10, 100),
    (2, 20, 200),
    (3, 30, 300),
    (4, 40, 400),
    (5, 50, 500),
    (6, 60, 600),
    (7, 70, 700),
    (8, 80, 800);

INSERT INTO tbl2 VALUES
    (10, 100, 1),
    (20, 200, 2),
    (30, 300, 1),
    (40, 400, 2),
    (50, 500, 6),
    (60, 600, 7),
    (70, 700, 8);

INSERT INTO tbl3 VALUES
    (10, 100, 3),
    (20, 200, 3),
    (30, 300, 4),
    (40, 400, 4),
    (50, 500, 5),
    (60, 600, 6),
    (70, 700, 7);

SELECT * FROM tbl1;

SELECT * FROM tbl2;

SELECT * FROM tbl3;

DELETE FROM tbl1 WHERE c1 = 2;

SELECT * FROM tbl1;

SELECT * FROM tbl2;

SELECT * FROM tbl3;

UPDATE tbl1 SET c1 = 10 WHERE c1 = 1;

SELECT * FROM tbl1;

SELECT * FROM tbl2;

SELECT * FROM tbl3;

UPDATE tbl2 SET c6 = 10 WHERE c4 <= 40;

SELECT * FROM tbl1;

SELECT * FROM tbl2;

SELECT * FROM tbl3;

{{Fail .Test}}
DELETE FROM tbl1 WHERE c1 = 10;

SELECT * FROM tbl1;

SELECT * FROM tbl2;

SELECT * FROM tbl3;

{{Fail .Test}}
DELETE FROM tbl1 WHERE c1 = 3;

UPDATE tbl1 SET c1 = 30 WHERE c1 = 3;

SELECT * FROM tbl1;

SELECT * FROM tbl2;

SELECT * FROM tbl3;

-- Drop table with foreign key reference first.
DROP TABLE IF EXISTS tbl3;

DROP TABLE IF EXISTS tbl2;

DROP TABLE IF EXISTS tbl1;
