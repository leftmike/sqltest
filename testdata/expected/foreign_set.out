--
-- Test FOREIGN KEY constraints
--
-- {{if eq Dialect "sqlite3"}}{{Skip}}{{end}}
DROP TABLE IF EXISTS tbl3;
DROP TABLE IF EXISTS tbl2;
DROP TABLE IF EXISTS tbl1;
CREATE TABLE tbl1 (
    c1 int PRIMARY KEY,
    c2 int,
    c3 int
);
CREATE TABLE tbl2 (
    c4 int PRIMARY KEY,
    c5 int,
    c6 int REFERENCES tbl1 ON DELETE SET DEFAULT ON UPDATE SET NULL DEFAULT 1
);
CREATE TABLE tbl3 (
    c4 int PRIMARY KEY,
    c5 int,
    c6 int REFERENCES tbl1 ON DELETE SET NULL ON UPDATE SET DEFAULT DEFAULT 4 NOT NULL
);
INSERT INTO tbl1 VALUES
    (1, 10, 100),
    (2, 20, 200),
    (3, 30, 300),
    (4, 40, 400),
    (5, 50, 500),
    (6, 60, 600),
    (7, 70, 700),
    (8, 80, 800);
INSERT INTO tbl2 VALUES
    (10, 100, 1),
    (20, 200, 2),
    (30, 300, 1),
    (40, 400, 2),
    (50, 500, 6),
    (60, 600, 7),
    (70, 700, 8);
INSERT INTO tbl3 VALUES
    (10, 100, 3),
    (20, 200, 3),
    (30, 300, 4),
    (40, 400, 4),
    (50, 500, 5),
    (60, 600, 6),
    (70, 700, 7);
SELECT * FROM tbl1;
   c1 c2  c3
   -- --  --
 1  1 10 100
 2  2 20 200
 3  3 30 300
 4  4 40 400
 5  5 50 500
 6  6 60 600
 7  7 70 700
 8  8 80 800
(8 rows)
SELECT * FROM tbl2;
   c4  c5 c6
   --  -- --
 1 10 100  1
 2 20 200  2
 3 30 300  1
 4 40 400  2
 5 50 500  6
 6 60 600  7
 7 70 700  8
(7 rows)
SELECT * FROM tbl3;
   c4  c5 c6
   --  -- --
 1 10 100  3
 2 20 200  3
 3 30 300  4
 4 40 400  4
 5 50 500  5
 6 60 600  6
 7 70 700  7
(7 rows)
DELETE FROM tbl1 WHERE c1 = 2;
SELECT * FROM tbl1;
   c1 c2  c3
   -- --  --
 1  1 10 100
 2  3 30 300
 3  4 40 400
 4  5 50 500
 5  6 60 600
 6  7 70 700
 7  8 80 800
(7 rows)
SELECT * FROM tbl2;
   c4  c5 c6
   --  -- --
 1 10 100  1
 2 20 200  1
 3 30 300  1
 4 40 400  1
 5 50 500  6
 6 60 600  7
 7 70 700  8
(7 rows)
SELECT * FROM tbl3;
   c4  c5 c6
   --  -- --
 1 10 100  3
 2 20 200  3
 3 30 300  4
 4 40 400  4
 5 50 500  5
 6 60 600  6
 7 70 700  7
(7 rows)
UPDATE tbl1 SET c1 = 10 WHERE c1 = 1;
SELECT * FROM tbl1;
   c1 c2  c3
   -- --  --
 1 10 10 100
 2  3 30 300
 3  4 40 400
 4  5 50 500
 5  6 60 600
 6  7 70 700
 7  8 80 800
(7 rows)
SELECT * FROM tbl2;
   c4  c5 c6
   --  -- --
 1 10 100   
 2 20 200   
 3 30 300   
 4 40 400   
 5 50 500  6
 6 60 600  7
 7 70 700  8
(7 rows)
SELECT * FROM tbl3;
   c4  c5 c6
   --  -- --
 1 10 100  3
 2 20 200  3
 3 30 300  4
 4 40 400  4
 5 50 500  5
 6 60 600  6
 7 70 700  7
(7 rows)
UPDATE tbl2 SET c6 = 10 WHERE c4 <= 40;
SELECT * FROM tbl1;
   c1 c2  c3
   -- --  --
 1 10 10 100
 2  3 30 300
 3  4 40 400
 4  5 50 500
 5  6 60 600
 6  7 70 700
 7  8 80 800
(7 rows)
SELECT * FROM tbl2;
   c4  c5 c6
   --  -- --
 1 10 100 10
 2 20 200 10
 3 30 300 10
 4 40 400 10
 5 50 500  6
 6 60 600  7
 7 70 700  8
(7 rows)
SELECT * FROM tbl3;
   c4  c5 c6
   --  -- --
 1 10 100  3
 2 20 200  3
 3 30 300  4
 4 40 400  4
 5 50 500  5
 6 60 600  6
 7 70 700  7
(7 rows)
{{Fail .Test}}
DELETE FROM tbl1 WHERE c1 = 10;
SELECT * FROM tbl1;
   c1 c2  c3
   -- --  --
 1 10 10 100
 2  3 30 300
 3  4 40 400
 4  5 50 500
 5  6 60 600
 6  7 70 700
 7  8 80 800
(7 rows)
SELECT * FROM tbl2;
   c4  c5 c6
   --  -- --
 1 10 100 10
 2 20 200 10
 3 30 300 10
 4 40 400 10
 5 50 500  6
 6 60 600  7
 7 70 700  8
(7 rows)
SELECT * FROM tbl3;
   c4  c5 c6
   --  -- --
 1 10 100  3
 2 20 200  3
 3 30 300  4
 4 40 400  4
 5 50 500  5
 6 60 600  6
 7 70 700  7
(7 rows)
{{Fail .Test}}
DELETE FROM tbl1 WHERE c1 = 3;
UPDATE tbl1 SET c1 = 30 WHERE c1 = 3;
SELECT * FROM tbl1;
   c1 c2  c3
   -- --  --
 1 10 10 100
 2 30 30 300
 3  4 40 400
 4  5 50 500
 5  6 60 600
 6  7 70 700
 7  8 80 800
(7 rows)
SELECT * FROM tbl2;
   c4  c5 c6
   --  -- --
 1 10 100 10
 2 20 200 10
 3 30 300 10
 4 40 400 10
 5 50 500  6
 6 60 600  7
 7 70 700  8
(7 rows)
SELECT * FROM tbl3;
   c4  c5 c6
   --  -- --
 1 10 100  4
 2 20 200  4
 3 30 300  4
 4 40 400  4
 5 50 500  5
 6 60 600  6
 7 70 700  7
(7 rows)
-- Drop table with foreign key reference first.
DROP TABLE IF EXISTS tbl3;
DROP TABLE IF EXISTS tbl2;
DROP TABLE IF EXISTS tbl1;
